packages/mui-system/src/createStyled.js

```diff
@@ -163,19 +178,117 @@ export default function createStyled(input = {}) {
             // On the server Emotion doesn't use React.forwardRef for creating components, so the created
             // component stays as a function. This condition makes sure that we do not interpolate functions
             // which are basically components used as a selectors.
-            return typeof stylesArg === 'function' && stylesArg.__emotion_real !== stylesArg
-              ? (props) => {
-                  return stylesArg({
-                    ...props,
-                    theme: resolveTheme({ ...props, defaultTheme, themeId }),
-                  });
+            if (typeof stylesArg === 'function' && stylesArg.__emotion_real !== stylesArg) {
+              return (props) => {
+                const resolvedStyles = stylesArg({
+                  ...props,
+                  theme: resolveTheme({ ...props, defaultTheme, themeId }),
+                });
+
+                let optionalVariants;
+                if (isPlainObject(resolvedStyles)) {
+                  if (resolvedStyles && resolvedStyles.variants) {
+                    optionalVariants = resolvedStyles.variants;
+                  }
+                  delete resolvedStyles.variants;
+                }
+                if (optionalVariants) {
+                  const variantsStyles = variantsResolver(
+                    props,
+                    transformVariants(optionalVariants),
+                    optionalVariants,
+                  );
+
+                  return [resolvedStyles, ...variantsStyles];
                 }
-              : stylesArg;
+
+                return resolvedStyles;
+              };
+            }
+            if (isPlainObject(stylesArg)) {
+              let transformedStylesArg = stylesArg;
+              let styledArgVariants;
+              if (stylesArg && stylesArg.variants) {
+                styledArgVariants = stylesArg.variants;
+              }
+              delete transformedStylesArg.variants;
+
+              if (styledArgVariants) {
+                transformedStylesArg = (props) => {
+                  let result = stylesArg;
+                  const variantStyles = variantsResolver(
+                    props,
+                    transformVariants(styledArgVariants),
+                    styledArgVariants,
+                  );
+                  variantStyles.forEach((variantStyle) => {
+                    result = deepmerge(result, variantStyle);
+                  });
+
+                  return result;
+                };
+              }
+              return transformedStylesArg;
+            }
+            return stylesArg;
           })
         : [];
 
       let transformedStyleArg = styleArg;
 
+      if (isPlainObject(styleArg)) {
```

Comment:
Adding the variants styles in case of one style arg.

Reply:
Logic here and the one at line 211 seem to be same. Can we call a scoped function at both places instead ?

Reply:
Nice catch. I thought I did it, but it was done only for the functions args :) It's done now

Reply:
It was not the same as in on the one we were modifying a object, but in the other we were returning a new function :D I just spend an hour figuring out what was broken :D